A58-client-side-weighted-round-robin-lb-policy.md

```diff
@@ -0,0 +1,251 @@
+
+In Java, the current subchannel list is updated directly as in `round_robin`. Upon getting an updated address list, it adds new subchannels to the subchannel list, starts watching the connectivity state of it, and requests connection. It also deletes removed subchannels from the subchannel list and shuts them down. It then updates the aggregate connectivity state, and if there is at least one `READY` subchannel, creates a new picker. The kind of picker created depends on the aggregate state: an active Picker (with a scheduler) if `READY`, EmptyPicker otherwise. When the connectivity state of a subchannel changes, it triggers updating the aggregate connectivity state.
+
+In Go, `baseBalancer` handles the address update as in `round_robin`. When the address set changes, the `baseBuilder` calls the picker builder of `weighted_round_robin` if the aggregate connectivity state is `READY`. When the connectivity state of a subchannel changes, it triggers the same process as when the address set changes.
+
+In all languages, when a subchannel state is updated to `IDLE`, the control plane immediately requests a connection and treats it as if it is in `CONNECTING` for the aggregation purpose.
+
+### Subchannel Weights
+
+To enforce `blackout_period` and `weight_expiration_period`, it tracks two timestamps along with the weight. `last_updated` tracks the last time a non-zero report was received. `non_empty_since` tracks when the first non-zero load report was received, and it is reset when a subchannel comes back to `READY` or the weight expires so that it can be updated with the next non-zero load report. The blackout and expiration are enforced when the weight value is looked up. When the weight expires, `non_empty_since` is reset.
+
+The weight of a backend is calculated as ***weight = qps / cpuutilization***.
+
+```
+function UpdateWeight(qps, cpu_utilization) {
+  var new_eight =
+     cpu_utilization == 0 ? 0 : qps / cpu_utilization;
+  if (new_eight == 0) return;
+  if (non_empty_since_ == infy) non_empty_since_ = now();
+  last_updated_ = now();
+  weight_ = new_weight;
+}
+function OnSubchannelBecomesReady() {non_empty_since_ = infy;}
+function GetWeight() {
+   if (now() - last_updated_ >= weight_expiration_period_) {
+      non_empty_since_ = infy;
+      return 0;
+   }
+   if (now() - non_empty_since_ < blackout_period_) return 0;
```

Comment:
Need to also check `blackout_period_ > 0`.

Reply:
Done.